### PR TITLE
feat: segregate logs

### PR DESCRIPTION
**Pull request by AbhishekA1509**
# Description
This PR adds:
- support to segregate logs in build history and deploy history.
- backward compatibility of old data.
- creation of intermediary state to organise data of logs in new logs.
- shifting of createMarkup outside p tag to avoid re-computations on each scroll.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] The title of the PR states what changed and the related issues number (used for the release note).
- [ ] Does this PR require documentation updates?
- [ ] I've updated documentation as required by this PR.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/Shared/Helpers.tsx [hunk 1]
```diff
@@ -707,3 +708,16 @@ export const decode = (data, isEncoded: boolean = false) =>
             agg[curr.key] = curr.value
             return agg
         }, {})
+
+export const getTimeDifference = (startTime: string, endTime: string): string => {
```
Comment (eshankvaish) [inline src/Shared/Helpers.tsx:712]: Please check for NaN

File: src/Shared/Components/CICDHistory/LogsRenderer.tsx [hunk 2]
```diff
@@ -149,43 +171,292 @@ export const LogsRenderer = ({
...
+            {} as Record<string, Record<string, StageDetailType>>,
+        )
+
+        // Map of stage as key and value as object with key as start time and value as boolean depicting if search key is present or not
+        const searchKeyStatusMap: Record<string, Record<string, boolean>> = {}
+
+        return streamDataList.reduce((acc, streamItem: string, index) => {
+            if (streamItem.startsWith(LOGS_STAGE_IDENTIFIER)) {
+                try {
+                    const { stage, startTime, endTime, status }: StageInfoDTO = JSON.parse(streamItem.split('|')[1])
```
Comment (eshankvaish) [inline src/Shared/Components/CICDHistory/LogsRenderer.tsx:274]: pls use a constant for |

File: src/Shared/Components/CICDHistory/LogsRenderer.tsx [hunk 3]
```diff
@@ -149,43 +171,292 @@ export const LogsRenderer = ({
...
+                    const existingStage = acc.find((item) => item.stage === stage && item.startTime === startTime)
+                    const previousExistingStage = previousStageMap[stage]?.[startTime]
+
```
Comment (eshankvaish) [inline src/Shared/Components/CICDHistory/LogsRenderer.tsx:277]:
```suggestion
                    const previousExistingStage = previousStageMap[stage]?.[startTime] ?? {}
```

File: src/Shared/Components/CICDHistory/LogsRenderer.tsx [hunk 4]
```diff
@@ -149,43 +171,292 @@ export const LogsRenderer = ({
...
+                } catch (e) {
+                    acc.push({
+                        stage: `Untitled stage ${index + 1}`,
+                        startTime: ZERO_TIME_STRING,
+                        endTime: ZERO_TIME_STRING,
+                        isOpen: false,
+                        logs: [],
+                        // Can not set status as FAILURE, as we will append latest logs to last stage
+                        status: StageStatusType.PROGRESSING,
+                    })
```
Comment (eshankvaish) [inline src/Shared/Components/CICDHistory/LogsRenderer.tsx:309]: these two can be commoned out

File: src/Shared/Components/CICDHistory/LogsRenderer.tsx [hunk 5]
```diff
@@ -149,43 +171,292 @@ export const LogsRenderer = ({
...
+            setLogsList(newLogs)
+            return
         }
+
+        const newStageList = getStageListFromStreamData()
+        setStageList(newStageList)
+        // NOTE: Not adding searchKey as dependency since on mount we would already have searchKey
+        // And for other cases we would use handleSearchEnter
+        // TODO: Check if stringifying the streamDataList is required
+    }, [JSON.stringify(streamDataList)])
```
Comment (eshankvaish) [inline src/Shared/Components/CICDHistory/LogsRenderer.tsx:351]: .

File: src/Shared/Components/CICDHistory/LogsRenderer.tsx [hunk 6]
```diff
@@ -149,43 +171,292 @@ export const LogsRenderer = ({
...
+                <div
+                    className="flexbox-col dc__gap-8"
+                    data-testid="check-logs-detail"
+                    style={{
+                        backgroundColor: '#0C1021',
+                    }}
+                >
+                    <div
+                        className="flexbox pl-12 logs-renderer__search-bar logs-renderer__filters-border-bottom dc__position-sticky dc__top-0 dc__zi-1"
+                        onKeyDown={stopPropagation}
```
Comment (eshankvaish) [inline src/Shared/Components/CICDHistory/LogsRenderer.tsx:386]: add a comment as to why this is required

**Review by eshankvaish — CHANGES_REQUESTED** [review]

File: src/Shared/Components/CICDHistory/LogsRenderer.tsx (hunk 4 above)
Comment (AbhishekA1509) [reply src/Shared/Components/CICDHistory/LogsRenderer.tsx:309]: Have removed acc.push in catch block no need of that

**Review by eshankvaish — DISMISSED** [review]

File: src/Shared/Components/CICDHistory/LogsRenderer.scss [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+                background: transparent;
+            }
+        }
+
+        .search-bar:focus-within {
+            border: none !important;
+        }
+    }
+
+    // TODO: Sync with product to replace this placeholder
```
Comment (vivek-devtron) [inline src/Shared/Components/CICDHistory/LogsRenderer.scss:27]: Please remove this if done

File: src/Shared/Components/CICDHistory/LogsRenderer.tsx [hunk 8]
```diff
@@ -149,43 +176,302 @@ export const LogsRenderer = ({
...
+        let isSearchKeyPresent = false
         try {
             // eslint-disable-next-line no-param-reassign
             log = log.replace(/\[[.]*m/, (m) => `\x1B[${m}m`)
+
+            if (targetSearchKey && areStagesAvailable) {
+                const logParts = log.split(ANSI_UP_REGEX)
+                const availableEscapeCodes = log.match(ANSI_UP_REGEX)
+                const parts = logParts.reduce((acc, part, index) => {
+                    try {
```
Comment (vivek-devtron) [inline src/Shared/Components/CICDHistory/LogsRenderer.tsx:199]: Please add comments

**Review by vivek-devtron — APPROVED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**